From 1a99c7aa3de08eb704cb136991873db97ccaa4a0 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Wed, 17 Nov 2021 13:37:47 +0900
Subject: [PATCH 146/196] Update coinbase addresses for Exodus hardfork (#232)

Coinbase payout addresses are going to update during the Exodus hardfork to new recipients.
---
 src/consensus/addresses_mainnet.h             |  31 +++-
 src/consensus/addresses_regtest.h             |  28 ++++
 src/consensus/addresses_testnet.h             |  28 ++++
 src/consensus/params.h                        |   1 +
 src/minerfund.cpp                             |  13 +-
 src/test/minerfund_tests.cpp                  |  35 ++++-
 test/functional/logos_feature_minerfund.py    |   4 +-
 .../logos_feature_minerfund_activation.py     | 132 ++++++++++++++++++
 8 files changed, 262 insertions(+), 10 deletions(-)
 create mode 100755 test/functional/logos_feature_minerfund_activation.py

diff --git a/src/consensus/addresses_mainnet.h b/src/consensus/addresses_mainnet.h
index 6d5fb9073..284836e57 100644
--- a/src/consensus/addresses_mainnet.h
+++ b/src/consensus/addresses_mainnet.h
@@ -33,6 +33,35 @@ namespace MainNet {
             "qp7s3y735fut4vn70tvjakyt6lww4lvr55ljvaxpyn",
             // Distro Bot
             "qp9cd8u62hzhqq7lz79aeqq3ssgfmyz03vc5c4g0ct",
-        }};
+        },
+        .exodus = {
+            // Aleph 0
+            "lotus_16PSJNf1EDEfGvaYzaXJCJZrXH4pgiTo7kyW61iGi",
+            // Maintainance
+            "lotus_16PSJNZhPZSyLeQaiEdy3a3gZdwQ2yAPqUm7WHRzN",
+            // Stamp
+            "lotus_16PSJHhULzYgupzGnvDysUFLxYxBVwx92JjXSyjNL",
+            // VN
+            "lotus_16PSJLMtCuFnyMb3rz4zbbhAtJgppp8GiY5NpGGev",
+            // Explorer
+            "lotus_16PSJJQGXSV1uMG2JHERQpVL3pVH7nwqu2YSiX96t",
+            // Distro
+            "lotus_16PSJHEXKQEL5M28Rc8rvYntymdKtPmVXJDEtbFLi",
+            // Evangelism
+            "lotus_16PSJNnGguQxEhHHW569L3g3MfhCNsS1Q6AWrJ8AH",
+            // Lotusia Citizen 1
+            "lotus_16PSJP5wAt1F45LycobZssN8Jk3TaR6QL9AXcRRty",
+            // Marketing
+            "lotus_16PSJHej3n56z43NC4PgYAAubk8opBXt7JXXfHjwc",
+            // Lotusia Citizen 2
+            "lotus_16PSJJSRTKjXKkvMtV6NX8SVEbTBwpjCWxQP7bHdi",
+            // Bots
+            "lotus_16PSJHuUsbDo5qZUjygQgtJLMsZgcpSB6wD459F6J",
+            // Bot
+            "lotus_16PSJJXVKnYb8xtUhHyBDSdd8P2xx6suUM7TaWBNx",
+            // Lotusia Citizen 3
+            "lotus_16PSJKHmQ2Lm8BNtnvPi1jC4rxJpCo8VqvJecj5Hb",
+        }
+    };
 } // namespace MainNet
 } // namespace RewardAddresses
diff --git a/src/consensus/addresses_regtest.h b/src/consensus/addresses_regtest.h
index 1af9a0fb4..0ab48dc10 100644
--- a/src/consensus/addresses_regtest.h
+++ b/src/consensus/addresses_regtest.h
@@ -33,6 +33,34 @@ namespace RegTest {
             "qp7s3y735fut4vn70tvjakyt6lww4lvr55ljvaxpyn",
             // Distro Bot
             "qp9cd8u62hzhqq7lz79aeqq3ssgfmyz03vc5c4g0ct",
+        },
+        .exodus = {
+            // Aleph 0
+            "lotus_16PSJNf1EDEfGvaYzaXJCJZrXH4pgiTo7kyW61iGi",
+            // Maintainance
+            "lotus_16PSJNZhPZSyLeQaiEdy3a3gZdwQ2yAPqUm7WHRzN",
+            // Stamp
+            "lotus_16PSJHhULzYgupzGnvDysUFLxYxBVwx92JjXSyjNL",
+            // VN
+            "lotus_16PSJLMtCuFnyMb3rz4zbbhAtJgppp8GiY5NpGGev",
+            // Explorer
+            "lotus_16PSJJQGXSV1uMG2JHERQpVL3pVH7nwqu2YSiX96t",
+            // Distro
+            "lotus_16PSJHEXKQEL5M28Rc8rvYntymdKtPmVXJDEtbFLi",
+            // Evangelism
+            "lotus_16PSJNnGguQxEhHHW569L3g3MfhCNsS1Q6AWrJ8AH",
+            // Lotusia Citizen 1
+            "lotus_16PSJP5wAt1F45LycobZssN8Jk3TaR6QL9AXcRRty",
+            // Marketing
+            "lotus_16PSJHej3n56z43NC4PgYAAubk8opBXt7JXXfHjwc",
+            // Lotusia Citizen 2
+            "lotus_16PSJJSRTKjXKkvMtV6NX8SVEbTBwpjCWxQP7bHdi",
+            // Bots
+            "lotus_16PSJHuUsbDo5qZUjygQgtJLMsZgcpSB6wD459F6J",
+            // Bot
+            "lotus_16PSJJXVKnYb8xtUhHyBDSdd8P2xx6suUM7TaWBNx",
+            // Lotusia Citizen 3
+            "lotus_16PSJKHmQ2Lm8BNtnvPi1jC4rxJpCo8VqvJecj5Hb",
         }};
 } // namespace RegTest
 } // namespace RewardAddresses
diff --git a/src/consensus/addresses_testnet.h b/src/consensus/addresses_testnet.h
index ae2c7d1a9..f319885ed 100644
--- a/src/consensus/addresses_testnet.h
+++ b/src/consensus/addresses_testnet.h
@@ -33,6 +33,34 @@ namespace TestNet {
             "qp7s3y735fut4vn70tvjakyt6lww4lvr55ljvaxpyn",
             // Distro Bot
             "qp9cd8u62hzhqq7lz79aeqq3ssgfmyz03vc5c4g0ct",
+        },
+        .exodus = {
+            // Aleph 0
+            "lotus_16PSJNf1EDEfGvaYzaXJCJZrXH4pgiTo7kyW61iGi",
+            // Maintainance
+            "lotus_16PSJNZhPZSyLeQaiEdy3a3gZdwQ2yAPqUm7WHRzN",
+            // Stamp
+            "lotus_16PSJHhULzYgupzGnvDysUFLxYxBVwx92JjXSyjNL",
+            // VN
+            "lotus_16PSJLMtCuFnyMb3rz4zbbhAtJgppp8GiY5NpGGev",
+            // Explorer
+            "lotus_16PSJJQGXSV1uMG2JHERQpVL3pVH7nwqu2YSiX96t",
+            // Distro
+            "lotus_16PSJHEXKQEL5M28Rc8rvYntymdKtPmVXJDEtbFLi",
+            // Evangelism
+            "lotus_16PSJNnGguQxEhHHW569L3g3MfhCNsS1Q6AWrJ8AH",
+            // Lotusia Citizen 1
+            "lotus_16PSJP5wAt1F45LycobZssN8Jk3TaR6QL9AXcRRty",
+            // Marketing
+            "lotus_16PSJHej3n56z43NC4PgYAAubk8opBXt7JXXfHjwc",
+            // Lotusia Citizen 2
+            "lotus_16PSJJSRTKjXKkvMtV6NX8SVEbTBwpjCWxQP7bHdi",
+            // Bots
+            "lotus_16PSJHuUsbDo5qZUjygQgtJLMsZgcpSB6wD459F6J",
+            // Bot
+            "lotus_16PSJJXVKnYb8xtUhHyBDSdd8P2xx6suUM7TaWBNx",
+            // Lotusia Citizen 3
+            "lotus_16PSJKHmQ2Lm8BNtnvPi1jC4rxJpCo8VqvJecj5Hb",
         }};
 } // namespace TestNet
 } // namespace RewardAddresses
diff --git a/src/consensus/params.h b/src/consensus/params.h
index edc280735..7dd47e0f1 100644
--- a/src/consensus/params.h
+++ b/src/consensus/params.h
@@ -16,6 +16,7 @@ namespace Consensus {
 
 struct CoinbaseAddresses {
     std::vector<std::string> genesis;
+    std::vector<std::string> exodus;
 };
 
 /**
diff --git a/src/minerfund.cpp b/src/minerfund.cpp
index df8721227..2977ad944 100644
--- a/src/minerfund.cpp
+++ b/src/minerfund.cpp
@@ -6,6 +6,7 @@
 
 #include <cashaddrenc.h> // For DecodeCashAddrContent
 #include <chainparams.h>
+#include <consensus/activation.h>
 #include <key_io.h> // For DecodeDestination
 
 static const CTxOut BuildOutput(const std::string &address,
@@ -22,6 +23,15 @@ static const CTxOut BuildOutput(const std::string &address,
     return {amount, script};
 }
 
+static const std::vector<std::string> &
+GetPayoutAddresses(const Consensus::Params &params,
+                   const CBlockIndex *pindexPrev) {
+    if (IsExodusEnabled(params, pindexPrev)) {
+        return params.coinbasePayoutAddresses.exodus;
+    }
+    return params.coinbasePayoutAddresses.genesis;
+}
+
 std::vector<CTxOut> GetMinerFundRequiredOutputs(const Consensus::Params &params,
                                                 const bool enableMinerFund,
                                                 const CBlockIndex *pindexPrev,
@@ -31,8 +41,7 @@ std::vector<CTxOut> GetMinerFundRequiredOutputs(const Consensus::Params &params,
     }
 
     const std::vector<std::string> &addresses =
-        params.coinbasePayoutAddresses.genesis;
-
+        GetPayoutAddresses(params, pindexPrev);
     const size_t numAddresses = addresses.size();
     std::vector<CTxOut> minerFundOutputs;
     const Amount shareAmount = blockReward / (int64_t(2 * numAddresses));
diff --git a/src/test/minerfund_tests.cpp b/src/test/minerfund_tests.cpp
index 9e279ce86..dd6a382dc 100644
--- a/src/test/minerfund_tests.cpp
+++ b/src/test/minerfund_tests.cpp
@@ -39,9 +39,9 @@
 BOOST_FIXTURE_TEST_SUITE(minerfund_tests, TestChain100Setup)
 
 static void
-CheckMinerfundOutputs(const std::vector<CTxOut> &outputs,
-                      const std::vector<std::string> &expectedAddresses,
-                      const Amount expectedAmount) {
+CheckPayoutOutputs(const std::vector<CTxOut> &outputs,
+                   const std::vector<std::string> &expectedAddresses,
+                   const Amount expectedAmount) {
     BOOST_CHECK_EQUAL(outputs.size(), expectedAddresses.size());
     const CChainParams mainNetParams =
         *CreateChainParams(CBaseChainParams::MAIN);
@@ -73,17 +73,32 @@ BOOST_AUTO_TEST_CASE(test_minerfund_required_outputs) {
     options.enableMinerFund = config.EnableMinerFund();
     BlockAssembler testAssembler(chainparams, *m_node.mempool, options);
 
+    const int64_t activationTime =
+        gArgs.GetArg("-exodusactivationtime", consensus.exodusActivationTime);
+
+    if (::ChainActive().Tip()->GetMedianTimePast() > activationTime) {
+        // This test doesn't work if -exodusactivationtime is in the past
+        return;
+    }
+
     // Simple consensus transaction.
     CScript scriptPubKey = CScript() << OP_RETURN;
     std::unique_ptr<CBlockTemplate> pblocktemplate;
 
     fCheckpointsEnabled = false;
 
-    for (size_t i = 0; i < 100; i++) {
+    // Some block num where the activation occurrs
+    size_t forkBlockNum = 50;
+
+    for (size_t blockNum = 0; blockNum < 100; blockNum++) {
         // Simple block creation
         BOOST_CHECK(pblocktemplate =
                         testAssembler.CreateNewBlock(scriptPubKey));
         CBlock *pblock = &pblocktemplate->block;
+        if (blockNum == forkBlockNum) {
+            SetMockTime(activationTime);
+            pblock->SetBlockTime(activationTime);
+        }
         const std::vector<CTxOut> &coinbaseOutputs = pblock->vtx[0]->vout;
         const CBlockIndex *pPrev = ::ChainActive().Tip();
 
@@ -99,8 +114,16 @@ BOOST_AUTO_TEST_CASE(test_minerfund_required_outputs) {
         BOOST_CHECK_EQUAL(coinbaseOutputs.size(), 15);
 
         // Check funding address outputs
-        CheckMinerfundOutputs(
-            outputs, consensus.coinbasePayoutAddresses.genesis, subsidy / 26);
+        // New addresses activate 6 blocks after the block time got bumped (MTP)
+        if (blockNum < forkBlockNum + 6) {
+            CheckPayoutOutputs(outputs,
+                               consensus.coinbasePayoutAddresses.genesis,
+                               subsidy / 26);
+        } else {
+            CheckPayoutOutputs(outputs,
+                               consensus.coinbasePayoutAddresses.exodus,
+                               subsidy / 26);
+        }
 
         // Update block so it can be connected, and the epoch will update
         // through this test.
diff --git a/test/functional/logos_feature_minerfund.py b/test/functional/logos_feature_minerfund.py
index 31378dd06..114af63b9 100755
--- a/test/functional/logos_feature_minerfund.py
+++ b/test/functional/logos_feature_minerfund.py
@@ -22,6 +22,7 @@ from test_framework.test_framework import BitcoinTestFramework
 from test_framework.txtools import pad_tx
 from test_framework.util import assert_equal
 
+ACTIVATION_TIME = 2000000000
 
 class MinerFundTest(BitcoinTestFramework):
     def set_test_params(self):
@@ -29,7 +30,8 @@ class MinerFundTest(BitcoinTestFramework):
         self.num_nodes = 2
         self.extra_args = [[
             '-enableminerfund',
-        ], ['-enableminerfund']]
+            f'-exodusactivationtime={ACTIVATION_TIME}',
+        ]] * 2
 
     def run_test(self):
         node = self.nodes[0]
diff --git a/test/functional/logos_feature_minerfund_activation.py b/test/functional/logos_feature_minerfund_activation.py
new file mode 100755
index 000000000..c368188e3
--- /dev/null
+++ b/test/functional/logos_feature_minerfund_activation.py
@@ -0,0 +1,132 @@
+#!/usr/bin/env python3
+# Copyright (c) 2021 The Logos Foundation
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+from decimal import Decimal
+
+from test_framework.blocktools import (
+    create_block,
+    create_coinbase,
+    prepare_block,
+    SUBSIDY,
+)
+from test_framework.messages import ToHex, CTxOut, CTxIn, COutPoint, CTransaction, COIN
+from test_framework.p2p import (
+    P2PDataStore,
+)
+from test_framework.script import (
+    CScript,
+    OP_HASH160,
+    OP_EQUAL,
+    OP_RETURN,
+)
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.txtools import pad_tx
+from test_framework.util import assert_equal
+
+ACTIVATION_TIME = 2000000000
+
+# see consensus/addresses_mainnet.h
+GENESIS_SCRIPTS = [
+    "a914b6c79031b71d86ab0d617e1e1e706ec4ee34b07f87",
+    "76a914b8ae1c47effb58f72f7bca819fe7fc252f9e852e88ac",
+    "76a914b50b86a893d80c9e2ee72b199612374b7b4c1cd888ac",
+    "76a914da76a31b6760dcb90aa469c15965da6e80096e4588ac",
+    "76a9141325d2d8ba6e8c7d99ff66d21530917cc73429d288ac",
+    "76a9146a171891ab9443020bd2755ef79c6e59efc5926588ac",
+    "76a91419e8d8d0edab6ec43f04b656bff72af78d63ff6588ac",
+    "76a914c6492d4e44dcd0051e60a8add6af02b2f291b2aa88ac",
+    "76a914b5aeafec9f2972110c4c6af9508a3c41e1d3c73b88ac",
+    "76a9147c28aa91b93faf8aee0a6520a0a83f42dbc4a45b88ac",
+    "76a914b18eb08c4978e73480743b1598061d3cf38e10a888ac",
+    "76a9147d0893d1a278bab27e7ad92ed88bd7dceafd83a588ac",
+    "76a9144b869f9a55c57003df178bdc801184109d904f8b88ac",
+]
+EXODUS_SCRIPTS = [
+    "76a914b50b86a893d80c9e2ee72b199612374b7b4c1cd888ac",
+    "76a914b20e7aad4df2a23bf412665019c41c96712acc6888ac",
+    "76a9141325d2d8ba6e8c7d99ff66d21530917cc73429d288ac",
+    "76a9146a171891ab9443020bd2755ef79c6e59efc5926588ac",
+    "76a9142a1e4c7ecaa87d49d2089a130413bd5cd2f99eaf88ac",
+    "76a91403f95bc854e8209bbd662c940326404be9bc8a8288ac",
+    "76a914b922f264776251674d69f13edbe2704377d8d4e688ac",
+    "76a914c3150272548e0bc82a54aaed8f634b3496bf34d088ac",
+    "76a9141199efaa97f8126b7f0a196dbefb000ac85bf16788ac",
+    "76a9142b54c6576b50bc8909201489ddc2d5c84317890a88ac",
+    "76a91419e8d8d0edab6ec43f04b656bff72af78d63ff6588ac",
+    "76a9142e2f1b2961a548b7107fdd4520c743afeabc991d88ac",
+    "76a914471d545492d9c94c1f630ba0507d3e378854f3d088ac",
+]
+
+class MinerFundActivationTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+        self.extra_args = [[
+            '-enableminerfund',
+            f'-exodusactivationtime={ACTIVATION_TIME}',
+        ]]
+
+    def run_test(self):
+        node = self.nodes[0]
+        node.add_p2p_connection(P2PDataStore())
+
+        # Mine a block, contains coinbase with miner fund
+        anyonecanspend_address = node.decodescript('51')['p2sh']
+        node.generatetoaddress(1, anyonecanspend_address)
+
+        def make_block_with_cb_scripts(scripts):
+            best_block_hash = node.getbestblockhash()
+            block_time = node.getblockheader(best_block_hash)['time']
+            block_height = node.getblockcount() + 1
+            coinbase = create_coinbase(block_height)
+            coinbase.vout[1].nValue = int(SUBSIDY * COIN) // 2
+            coinbase.vout[1].scriptPubKey = CScript([OP_HASH160, bytes(20), OP_EQUAL])
+            for script in scripts:
+                coinbase.vout.append(CTxOut(int(SUBSIDY * COIN) // 26, CScript(bytes.fromhex(script))))
+            coinbase.rehash()
+            block = create_block(int(best_block_hash, 16),
+                                coinbase, block_height, block_time + 1)
+            return block
+
+        # Pre-fork minerfund with wrong (i.e. post-fork) scripts
+        block = make_block_with_cb_scripts(EXODUS_SCRIPTS)
+        prepare_block(block)
+        assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+
+        # With correct scripts, except last script
+        block = make_block_with_cb_scripts(
+            GENESIS_SCRIPTS[:-1] + [EXODUS_SCRIPTS[-1]])
+        prepare_block(block)
+        assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+
+        # With correct scripts
+        block = make_block_with_cb_scripts(GENESIS_SCRIPTS)
+        prepare_block(block)
+        assert_equal(node.submitblock(ToHex(block)), None)
+
+        # Set mocktime
+        node.setmocktime(ACTIVATION_TIME)
+
+        # Mine 11 blocks with ACTIVATION_TIME in the middle
+        # That moves MTP exactly to ACTIVATION_TIME
+        for i in range(-6, 6):
+            block = make_block_with_cb_scripts(GENESIS_SCRIPTS)
+            block.nTime = ACTIVATION_TIME + i
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+
+        # Now the using the genesis addresses fails
+        block = make_block_with_cb_scripts(GENESIS_SCRIPTS)
+        prepare_block(block)
+        assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+
+        # Using new scripts now works
+        block = make_block_with_cb_scripts(EXODUS_SCRIPTS)
+        prepare_block(block)
+        assert_equal(node.submitblock(ToHex(block)), None)
+
+
+if __name__ == '__main__':
+    MinerFundActivationTest().main()
-- 
2.39.5 (Apple Git-154)

